raft.c

```diff
@@ -2321,31 +2329,30 @@ void handleDebug(RedisRaftCtx *rr, RaftReq *req)
  */
 void applyShardGroupChange(RedisRaftCtx *rr, raft_entry_t *entry)
 {
-    ShardGroup sg;
+    ShardGroup * sg = ShardGroupCreate();
     RRStatus ret;
 
-    if (ShardGroupDeserialize(entry->data, entry->data_len, &sg) != RR_OK) {
+    if (ShardGroupDeserialize(entry->data, entry->data_len, sg) != RR_OK) {
         LOG_ERROR("Failed to deserialize ADD_SHARDGROUP payload: [%.*s]",
                 entry->data_len, entry->data);
         return;
     }
 
     switch (entry->type) {
         case RAFT_LOGTYPE_ADD_SHARDGROUP:
-            if ((ret = ShardingInfoAddShardGroup(rr, &sg)) != RR_OK)
+            if ((ret = ShardingInfoAddShardGroup(rr, sg)) != RR_OK)
```

Comment:
Should we call `ShardGroupFree()` on failure?

Reply:
Hmm, maybe not. If I'm not mistaken, `ShardingInfoAddShardGroup()` destroys the object in some cases itself but not in all cases. Is it possible to revert `sg` to be stack variable and and let these functions create/copy ShardGroup object itself if needed. It's just not obvious who is the owner of the object, where should we destroy the object. Maybe we should you take another look at this.

It'll be nice to have a test case that hits these parts just to catch leak. (If we don't have already and easy to add one)

Reply:
so, I figured out how to test this, needs some changes to applyShardGroupChange to change correctly, working on it.